projects/plugins/crm/src/automation/commons/conditions/contacts/class-contact-tag.php

```diff
@@ -0,0 +1,209 @@
+<?php
+/**
+ * Jetpack CRM Automation Contact_Tag condition.
+ *
+ * @package automattic/jetpack-crm
+ */
+
+namespace Automattic\Jetpack\CRM\Automation\Conditions;
+
+use Automattic\Jetpack\CRM\Automation\Attribute_Definition;
+use Automattic\Jetpack\CRM\Automation\Automation_Exception;
+use Automattic\Jetpack\CRM\Automation\Base_Condition;
+use Automattic\Jetpack\CRM\Automation\Data_Types\Data_Type_Contact;
+
+/**
+ * Contact_Tag condition class.
+ *
+ * @since $$next-version$$
+ */
+class Contact_Tag extends Base_Condition {
+
+	/**
+	 * Contact_Tag constructor.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param array $step_data The step data.
+	 */
+	public function __construct( array $step_data ) {
+		parent::__construct( $step_data );
+
+		$this->valid_operators = array(
+			'tag_added'   => __( 'Tag was added', 'zero-bs-crm' ),
+			'tag_removed' => __( 'Tag was removed', 'zero-bs-crm' ),
+			'has_tag'     => __( 'Has tag', 'zero-bs-crm' ),
+		);
+
+		$this->set_attribute_definitions(
+			array(
+				new Attribute_Definition( 'operator', __( 'Operator', 'zero-bs-crm' ), __( 'Determines how the field is compared to the specified value.', 'zero-bs-crm' ), Attribute_Definition::SELECT, $this->valid_operators ),
+				new Attribute_Definition( 'tag', __( 'Tag', 'zero-bs-crm' ), __( 'Contact Tag to compare with.', 'zero-bs-crm' ), Attribute_Definition::TEXT ),
+			)
+		);
+	}
+
+	/**
+	 * Checks if a given tag name exists in the tags array.
+	 *
+	 * @param array  $data     The data array containing the 'tags' key.
+	 * @param string $tag_name The name of the tag to check for.
+	 *
+	 * @return bool True if the tag name exists, false otherwise.
+	 */
+	private function has_tag_by_name( $data, $tag_name ) {
+		if ( ! isset( $data['tags'] ) || ! is_array( $data['tags'] ) ) {
+			return false;
+		}
+
+		foreach ( $data['tags'] as $tag ) {
+			if ( isset( $tag['name'] ) && $tag['name'] === $tag_name ) {
+					return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Executes the condition. If the condition is met, the value stored in the
+	 * attribute $condition_met is set to true; otherwise, it is set to false.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param mixed  $data Data passed from the trigger.
+	 * @param ?mixed $previous_data (Optional) The data before being changed.
+	 * @return void
+	 *
+	 * @throws Automation_Exception If an invalid operator is encountered.
+	 */
+	public function execute( $data, $previous_data = null ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( $previous_data === null ) {
+			$this->logger->log( 'Invalid previous contact tag data' );
```

Comment:
I'm not sure if this is okay for all the cases, for instance, `has_tag` condition should work with new contacts. For example, when a new Contact is created, we don't have previous data, it will be null, but we can check `has_tag` condition.

On the other hand, for `tag_added` or `tag_removed` we need information about the previous state.

Reply:
In summary, that check would make sense only if the condition is for a tag added or removed.

Reply:
Addressed in https://github.com/Automattic/jetpack/pull/32776

Reply:
I suppose it is this commit: https://github.com/Automattic/jetpack/pull/32776/commits/7f5b2250b2edf0d9edf6719c9a14cc43f07f69b2

Reply:
👍  Nice extraction!